### DISCUSSION TITLE
Generic Exception Handler to capture all types of exception.

### DISCUSSION THREAD

**Discussion by Babak-abd** [fastapi/fastapi#9478] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from fastapi import FastAPI, Request, HTTPException
from starlette.exceptions import HTTPException as StarletteHTTPException
from fastapi.responses import PlainTextResponse, JSONResponse
from fastapi.encoders import jsonable_encoder

from uvicorn.main import run

app = FastAPI()


'''
@app.exception_handler(StarletteHTTPException)
async def http_exception_handler(request, exc):
    return PlainTextResponse(str(exc.detail), status_code=exc.status_code)
'''

# General exception handler
@app.exception_handler(Exception)
async def general_exception_handler(request: Request, exc: Exception):
    return JSONResponse(
        status_code= exec.status_code if isinstance(exc, HTTPException)  else 500,
        content=jsonable_encoder({"error:": str(exc)}),
    )


@app.get('/api/v1/example', tags=["metrics"])
async def metrics():
    #raise Exception("Not Acceptable")
    raise HTTPException(status_code=406, detail="Not Acceptable")


# Start the app
if __name__ == '__main__':
    #can provide appropriate ssl params (as defined in the uvicorn run method).
    run(app, host='0.0.0.0', port=8000)
```

### Description

I am trying to add a single exception handler that could handle all types of exceptions.

My main question is why in my snippet code, Raising a HTTPException will not be handled in my general_exception_handler ( it will work with http_exception_handler)

I checked HTTPException is derived from  StarletteHTTPException which is also drived from Exception so I was hoping that this should work !!

### Operating System

Windows

### FastAPI Version

0.95.1

### Python Version

Python 3.11.3

**Comment by jgould22** [comment, marked as answer]
That is just not how it works.

The default exception handler middleware is still in place and it checks the type. 

https://github.com/encode/starlette/blob/7c4fd9cbf04ecd57e82fe4761f8d7976a9cb53bd/starlette/middleware/exceptions.py#L67

**Reply by muhammadnasr** [reply]
So how can I have a single exception handler for everything?

**Reply by obscurebyron** [reply]
I would also like to know this.

**Reply by whokc-github** [reply]
Hi, please check @SGobet reply below. He added a header line to get through the CORS filter. I added it and it works right away.

**Comment by HenriBlacksmith** [comment]
From what I see in the Starlette code, when you register an exception handler for the `Exception` class or `500` HTTP code, it is assigned to the `ServerErrorMiddleware` instead of `ExceptionMiddleware`.

https://github.com/encode/starlette/blob/7c4fd9cbf04ecd57e82fe4761f8d7976a9cb53bd/starlette/applications.py#L90

For some reason in my case using FastAPI 0.100.0 registering a generic error handler for `Exception` has no effect (I found other threads around this in other versions).

Still reading the code, looks like the handler should be used if it is set and if debug mode is not enabled: https://github.com/encode/starlette/blob/7c4fd9cbf04ecd57e82fe4761f8d7976a9cb53bd/starlette/middleware/errors.py#L173

It might be related to the `response_started` value in my custom handler: https://github.com/encode/starlette/blob/7c4fd9cbf04ecd57e82fe4761f8d7976a9cb53bd/starlette/middleware/errors.py#L178C20-L178C36

**Comment by yanggeorge** [comment]
This problem has been bothering me for a long time.

**Comment by SGobet** [comment]
Whats the recommended approach to have a generic exception handler?

**Reply by SGobet** [reply]
We ran into very odd issues mixing a generic exception handler `@app.exception_handler(Exception)`

```python
@app.exception_handler(Exception)
def exception_handler(_: Request, exc: Exception):
    return JSONResponse(
        status_code=500,
        content={
            "detail": str(exc),
            "error_type": type(exc).__name__,
        },
        headers={
            "Access-Control-Allow-Origin": "*"
        }
    )
```

We also have a cors handler that's very open
```CORSMiddleware, allow_origins=["*"], allow_methods=["*"], allow_headers=["*"], allow_credentials=True```

Any request handled by the generic Exception handler does not seem to have cors headers applied.
Required us to manually set the access control headers as shown above

**Reply by whokc-github** [reply]
Hello SGobet, thanks for posting this. I am just missing this header, has been searching for a week. Its a life-saver. Bravo!

**Comment by khalo-sa** [comment]
Thanks to @SGobet's [comment](https://github.com/fastapi/fastapi/discussions/9478#discussioncomment-10779142) I realized the missing CORS headers were responsible for the very inconsistent behaviour in my frontend when using a generic exception handler. I came up with the following minimal code comparing two ways of setting up CORS in a Fastapi app: 

TLDR: use the "wrap" strategy and you should be good!

```python
from typing import Literal

from fastapi import FastAPI, HTTPException, Request
from fastapi.responses import JSONResponse
from fastapi.testclient import TestClient
from starlette.middleware.cors import CORSMiddleware

ORIGIN = "http://testclient"
CORS_TYPE = Literal["none", "default", "wrap"]


class AppException(Exception):
    pass


class AppExceptionFoo(AppException):
    pass


class AppExceptionBar(AppException):
    pass


def create_app(cors: CORS_TYPE):
    app = FastAPI()

    @app.get("/")
    async def root():
        return {"content": "ok"}

    @app.get("/foo")
    async def foo():
        raise HTTPException(status_code=500, detail="blub")

    @app.get("/bar")
    async def bar():
        raise AppExceptionFoo()

    @app.get("/baz")
    async def baz():
        raise AppExceptionBar()

    @app.exception_handler(Exception)
    async def exception_handler(request: Request, exc: Exception):
        match exc:
            case AppExceptionFoo():
                status = 400
                detail = "Bad request"
            case AppExceptionBar():
                status = 401
                detail = "Unauthorized"
            case _:
                status = 500
                detail = "An unexpected error occurred"

        return JSONResponse(
            content={
                "status": status,
                "detail": detail,
            },
            status_code=status,
        )

    match cors:
        case "none":
            pass
        case "default":
            # https://fastapi.tiangolo.com/tutorial/cors/#use-corsmiddleware
            app.add_middleware(
                CORSMiddleware,
                allow_origins=[ORIGIN],
                allow_credentials=True,
                allow_methods=["*"],
                allow_headers=["*"],
            )
        case "wrap":
            # https://github.com/encode/starlette/issues/617#issuecomment-741876645
            app = CORSMiddleware(
                app,
                allow_origins=[ORIGIN],
                allow_credentials=True,
                allow_methods=["*"],
                allow_headers=["*"],
            )

    return app


if __name__ == "__main__":
    cors_: list[CORS_TYPE] = ["none", "default", "wrap"]
    for cors in cors_:
        print()
        print(f"--- Testing with cors={cors} ---")
        print()
        app = create_app(cors)
        client = TestClient(
            app,
            raise_server_exceptions=False,
            headers={"Origin": ORIGIN},
        )

        for endpoint in ["/", "/foo", "/bar", "/baz"]:
            print(endpoint)
            response = client.get(endpoint)
            print("status_code: ", response.status_code)
            print("json: ", response.json())
            print("cors headers: ", response.headers.get("Access-Control-Allow-Origin"))
            print()

        print()
```

If you run this, it prints:

```
--- Testing with cors=none ---

/
status_code:  200
json:  {'content': 'ok'}
cors headers:  None

/foo
status_code:  500
json:  {'detail': 'blub'}
cors headers:  None

/bar
status_code:  400
json:  {'status': 400, 'detail': 'Bad request'}
cors headers:  None

/baz
status_code:  401
json:  {'status': 401, 'detail': 'Unauthorized'}
cors headers:  None



--- Testing with cors=default ---

/
status_code:  200
json:  {'content': 'ok'}
cors headers:  http://testclient

/foo
status_code:  500
json:  {'detail': 'blub'}
cors headers:  http://testclient

/bar
status_code:  400
json:  {'status': 400, 'detail': 'Bad request'}
cors headers:  None

/baz
status_code:  401
json:  {'status': 401, 'detail': 'Unauthorized'}
cors headers:  None



--- Testing with cors=wrap ---

/
status_code:  200
json:  {'content': 'ok'}
cors headers:  http://testclient

/foo
status_code:  500
json:  {'detail': 'blub'}
cors headers:  http://testclient

/bar
status_code:  400
json:  {'status': 400, 'detail': 'Bad request'}
cors headers:  http://testclient

/baz
status_code:  401
json:  {'status': 401, 'detail': 'Unauthorized'}
cors headers:  http://testclient
```

**Comment by beck3905** [comment]
I'm just running into this issue now. I'm not doing anything with CORS in my API right now. If I have my own `ApplicationException` type and setup the exception handler for that type then even with a 500 (HTTPException) error I get a 

```json
{
    "detail": "Internal Server Error"
}
```
but if I change the handler to capture `Exception` then I don't get a JSON response and the response body is just
```
Internal Server Error
```

**Comment by cameronlund4** [comment]
The behavior @HenriBlacksmith mentioned was causing some weirdness.

FastAPI's [`@app.exception_handler(...)`](https://github.com/fastapi/fastapi/blob/f23550223475e0e173ef63b0d54236fc105a8867/fastapi/applications.py#L4571) calls Starlette's [`add_exception_handler(...)`](https://github.com/Kludex/starlette/blob/main/starlette/applications.py#L134), which registers it to a dictionary of `Exception` to handler:
```python
def add_exception_handler(
    self,
    exc_class_or_status_code: int | type[Exception],
    handler: ExceptionHandler,
) -> None:  # pragma: no cover
    self.exception_handlers[exc_class_or_status_code] = handler
```

Then, when Starlette [goes to build the middleware stack](https://github.com/Kludex/starlette/blob/main/starlette/applications.py#L80), it handles the key `Exception` and `500` separately:
```python
    def build_middleware_stack(self) -> ASGIApp:
        debug = self.debug
        error_handler = None
        exception_handlers: dict[Any, ExceptionHandler] = {}

        for key, value in self.exception_handlers.items():
            if key in (500, Exception):
                error_handler = value # ‼️ If it is 500 or Exception, we set this separate variable
            else:
                exception_handlers[key] = value # ‼️ Normal exceptions get registered here

        middleware = (
            [Middleware(ServerErrorMiddleware, handler=error_handler, debug=debug)] # ‼️ 500 or Exception
            + self.user_middleware
            + [Middleware(ExceptionMiddleware, handlers=exception_handlers, debug=debug)] # ‼️ Everything else
        )

        app = self.router
        for cls, args, kwargs in reversed(middleware):
            app = cls(app, *args, **kwargs)
        return app
```
This `ServerErrorMiddleware` gets run _after_ all your other exception handlers run, and your own middleware. I believe this is why people were having issues with CORS above. The exception bubbles through the CORS middleware, and then the response is generated in the `ServerErrorMiddleware`. Since this is after the CORS middleware can run, there is no header added.

It also has some quirks relative to the `ExceptionMiddleware`, such as [always re-raising your exception regardless of your handler](https://github.com/Kludex/starlette/blob/main/starlette/middleware/errors.py#L183). In my case, I believe this was causing gunicorn errors and creating a lot of noise.

The workaround I found was to manually register an `ExceptionHandler` as middleware in my app, therefore it runs in the `self.user_middleware` section of the above snippet, and properly allows me to catch generic exceptions in the same way a normal `@app.exception_handler(...)` does. This way, your generated response still runs through your middleware as normal:
```python
from starlette.middleware.exceptions import ExceptionMiddleware

def uncaught_exception_handler(exception: Exception) -> JSONResponse:
    ...

app.add_middleware(
    ExceptionMiddleware, handlers={Exception: uncaught_exception_handler}, debug=False
)
```
Note, if your middleware raises an exception, it will now not go through this handler. So, depending on your case, there could be value in also having the standard `@app.exception_handler(Exception)`.
